core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -78,34 +77,39 @@ public class AndroidPeripheral internal constructor(
         combine(_ready, state) { ready, state -> ready && state == State.Connected }.first { it }
     }
 
+    private fun establishConnection(): Connection =
+        bluetoothDevice.connect(
+            context,
+            _state,
+            invokeOnClose = { connectJob.value = null }
+        ) ?: throw ConnectionRejectedException()
+
+    /** Creates a connect [Job] that completes when connection is established, or failure occurs. */
     private fun createConnectJob(): Job = scope.launch(start = LAZY) {
         _ready.value = false
 
-        val connection =
-            bluetoothDevice.connect(context, _state) ?: throw ConnectionRejectedException()
-        _connection = connection
+        val connection = establishConnection().also { _connection = it }
+        connection
+            .characteristicChanges
+            .onEach(observers.characteristicChanges::emit)
+            .launchIn(scope, start = UNDISPATCHED)
 
         try {
-            connection
-                .characteristicChanges
-                .onEach(observers.characteristicChanges::emit)
-                .catch { cause -> if (cause !is ConnectionLostException) throw cause }
-                .launchIn(scope, start = UNDISPATCHED)
-
             suspendUntilConnected()
             discoverServices()
             observers.rewire()
         } catch (t: Throwable) {
-            connection.close()
-            _connection = null
-            if (t !is ConnectionLostException) throw t
+            dispose()
+            throw t
         }
 
         println("Ready")
         _ready.value = true
-    }.apply {
-        // fixme: Clear on disconnect (not at end of connect).
-        invokeOnCompletion { connectJob.value = null }
+    }
+
+    private fun dispose() {
```

Comment:
Should this function name be a little more verbose to point out that it's the connection being disposed? Looking at this from the outside, I would have guessed that this function would be cleaning up some state for Peripheral at large--not just the Peripheral's connection.

Reply:
Technically, the `Connection` object internally holds either a dispatcher or thread that needs to be cleaned up (and is done via callback from the `Connection.close`). So maybe `Connection.close` isn't verbose enough? Happy to change any names that make things more readable, not sure what to change it to though. Also not as important for `private` functions. 🤷